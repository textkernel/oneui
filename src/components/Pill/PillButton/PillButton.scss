@use 'sass:math';
@import '../../../themes/oneui/oneui';

$pill-height: 32px !default;
$pill-label-height: 17px !default;
$pill-button-size: 20px !default;

$half-pill-height: math.div($pill-height, 2);

@keyframes labelIn {
    from {
        opacity: 0;
        transform: translateY(5px);
    }
    to {
        opacity: 1;
        transform: translateY(0);
    }
}

.PillButton {
    display: inline-flex;
    flex-direction: column;
    justify-items: start;

    &__label {
        box-sizing: border-box;
        height: $pill-label-height;
        padding: 0 $half-pill-height;
        font-size: var(--font-size-normal);
        text-transform: uppercase;
        color: var(--color-brand-50);

        &--isActive {
            animation: labelIn var(--transition-duration) forwards;
            animation-timing-function: ease-out;
        }
    }

    &__pillLabel {
        display: inline-block;
        padding-bottom: 1px;
    }

    &__pill {
        display: flex;
        justify-content: space-between;
        align-items: center;
        box-sizing: border-box;
        height: $pill-height;
        border-radius: $half-pill-height;
        border: 1px solid var(--color-neutral-30);
        padding-right: var(--space-75);
        padding-left: $half-pill-height;
        color: var(--color-neutral-70);
        cursor: pointer;
        transition: all var(--transition-duration) ease;

        &:focus,
        &:hover {
            outline: none;
            border-color: var(--color-neutral-80);
            background-color: var(--color-neutral-10);
        }

        &--isOpen {
            &:hover {
                background-color: var(--color-neutral-10);
            }
        }

        &--isActive {
            border: 1px solid var(--color-brand-50);
            background-color: var(--color-brand-50);
            color: var(--color-background);

            &:focus,
            &:hover {
                background-color: var(--color-brand-60);
                border-color: var(--color-brand-60);
            }
        }
    }

    &__button {
        display: flex;
        justify-content: center;
        align-items: center;
        position: relative;
        left: 1px;
        padding: 2px;
        font-size: 16px;
        border: 1px solid transparent;
        border-radius: 50%;
        margin-left: var(--space-75);
        background-color: transparent;
        color: inherit;
        cursor: pointer;
        box-sizing: content-box;
        transition: background-color var(--transition-duration) ease;

        &--isOpen {
            transform: rotate(180deg);

            &:hover {
                background-color: transparent;
            }
        }
    }

    &__arrowIcon {
        transform: translateY(1px);

        &--isOpen {
            transform: translateY(0px);
        }
    }
}
