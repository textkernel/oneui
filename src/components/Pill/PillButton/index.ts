export { PillButton, Props as PillButtonProps } from './PillButton';
