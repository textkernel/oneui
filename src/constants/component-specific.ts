/** Constants that are component specific but do not merit a separate file */

/** Autocomplete */
export const NUMBER_OF_SUGGESTION_LOADING_PLACEHOLDERS = 5;

/** Button */
export const BUTTON_TYPES = ['submit', 'button'] as const;

export type ButtonType = typeof BUTTON_TYPES[number];

/** Input */
export const INPUT_TYPES = [
    'text',
    'password',
    'email',
    'date',
    'datetime-local',
    'month',
    'tel',
    'number',
    'url',
    'search',
    'time',
    'week',
    'color',
] as const;

export type InputType = typeof INPUT_TYPES[number];

/** List */
export const LIST_NAVIGATION_DIRECTIONS = {
    UP: 'ArrowUp',
    DOWN: 'ArrowDown',
};

/** PopupBase, MultiActionButton */
export const POPUP_PLACEMENTS = [
    'auto',
    'auto-start',
    'auto-end',
    'top',
    'top-start',
    'top-end',
    'right',
    'right-start',
    'right-end',
    'bottom',
    'bottom-start',
    'bottom-end',
    'left',
    'left-start',
    'left-end',
] as const;

export type PopupPlacement = typeof POPUP_PLACEMENTS[number];

/** Gauge */
export const GAUGE_RADIUS = 100;
