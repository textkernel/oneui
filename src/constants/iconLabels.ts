export const CLOSE_ICON_LABEL = 'X';
export const UP_ARROW_LABEL = 'arrow up';
export const DOWN_ARROW_LABEL = 'arrow down';
