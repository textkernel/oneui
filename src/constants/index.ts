export * from './style-related';
export * from './component-specific';
export * from './keyboard';
export * from './characters';
