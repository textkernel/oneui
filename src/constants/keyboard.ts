export const ENTER_KEY = 'Enter';
export const BACKSPACE_KEY = 'Backspace';
export const ESCAPE_KEY = 'Escape';
export const TAB_KEY = 'Tab';
export const SPACE_KEY = ' ';
