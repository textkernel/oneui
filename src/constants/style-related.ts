export const VALIDATION_CONTEXTS = ['success', 'warning', 'danger'] as const;

export const CONTEXTS = ['info', ...VALIDATION_CONTEXTS] as const;

export const SIZES = ['small', 'normal', 'large'] as const;
export const HEADING_SIZES = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'] as const;
export const CHECKBOX_VIEWBOX = [0, 0, 10, 10] as const;
export const COLORS = [
    '#AFF5E2',
    '#D6F481',
    '#FFF16B',
    '#FFD571',
    '#FFC4A6',
    '#E5DAC9',
    '#ECCBFF',
    '#FFC5EB',
    '#FCC1CF',
    '#C4D7FC',
    '#ADDDFF',
    '#75DFFB',
];

export type ValidationContext = (typeof VALIDATION_CONTEXTS)[number];
export type Context = (typeof CONTEXTS)[number];
export type Size = (typeof SIZES)[number];
