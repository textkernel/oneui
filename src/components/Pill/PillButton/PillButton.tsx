import * as React from 'react';
import { IoIosArrowDown, IoIosClose } from 'react-icons/io';
import { SingleReactNode } from '../../../customTypes/types';
import { bem } from '../../../utils';
import { ENTER_KEY } from '../../../constants';
import styles from './PillButton.scss';

export interface Props {
    /** Wether the dropdown is open or closed */
    isOpen?: boolean;
    /** If pill is in default state, meaning it has content but cannot be reset. */
    isContentDefault?: boolean;
    /** a function to be called when dropdown should be toggled */
    toggleDropdown: () => void;
    /** a function to be called to clear the pill/filter content */
    onClear: () => void;
    /** name describing the pill/filter */
    name: string;
    /** label describing the content of an active filter/pill */
    content?: SingleReactNode;
}

const { block, elem } = bem('PillButton', styles);

export const PillButton = React.forwardRef<HTMLElement, Props>(
    (
        {
            isOpen = false,
            isContentDefault = false,
            toggleDropdown,
            onClear,
            name,
            content = null,
            ...rest
        },
        ref
    ) => {
        const isActive = !!content;
        const propsForBem = { isOpen, isActive };

        const labelRef = React.createRef();
        const pillRef = React.createRef();
        const [pillMinWidth, setPillMinWidth] = React.useState(0);

        // make sure label is not longer then pill
        React.useLayoutEffect(() => {
            if (isActive) {
                const labelEl = labelRef.current as HTMLElement;
                const { width: labelWidth } = labelEl.getBoundingClientRect();
                const pillEl = pillRef.current as HTMLElement;
                const { width: pillWidth } = pillEl.getBoundingClientRect();

                if (labelWidth > pillWidth) {
                    setPillMinWidth(labelWidth);
                }
            }
        }, [isActive, labelRef, pillMinWidth, pillRef]);

        let buttonIcon = <IoIosArrowDown {...elem('arrowIcon', propsForBem)} />;
        let isButtonClickable = false;

        if (isActive && !isOpen && !isContentDefault) {
            buttonIcon = <IoIosClose size={22} />;
            isButtonClickable = true;
        }

        const buttonClick = isButtonClickable
            ? (e) => {
                  e.stopPropagation();
                  e.preventDefault();
                  onClear();
              }
            : undefined;

        const handleKeyDownOnButton = isButtonClickable
            ? (e) => {
                  if (e.key === ENTER_KEY) {
                      if (buttonClick) {
                          buttonClick(e);
                      }
                  }
              }
            : undefined;

        const handleKeyDownOnPill = (e) => {
            if (e.key === ENTER_KEY) {
                e.preventDefault();
                toggleDropdown();
            }
        };

        return (
            <div ref={ref} {...rest} {...block({ ...propsForBem, ...rest })}>
                <div ref={labelRef} {...elem('label', propsForBem)}>
                    {isActive && name}
                </div>
                <div
                    ref={pillRef}
                    {...elem('pill', propsForBem)}
                    style={pillMinWidth ? { minWidth: pillMinWidth } : undefined}
                    onClick={toggleDropdown}
                    onKeyDown={handleKeyDownOnPill}
                    tabIndex="0"
                    role="button"
                >
                    <span {...elem('pillLabel', propsForBem)}>{content || name}</span>
                    <button
                        type="button"
                        {...elem('button', propsForBem)}
                        onClick={buttonClick}
                        onKeyDown={handleKeyDownOnButton}
                    >
                        {buttonIcon}
                    </button>
                </div>
            </div>
        );
    }
);

PillButton.displayName = 'PillButton';
