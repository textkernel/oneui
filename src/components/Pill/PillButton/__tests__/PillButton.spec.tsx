import React from 'react';
import toJson from 'enzyme-to-json';
import { PillButton } from '../PillButton';
import { ENTER_KEY } from '../../../../constants';

describe('<PillButton> component', () => {
    const toggleDropdownMock = jest.fn();
    const onClearMock = jest.fn();
    const name = 'Pill name';
    const content = 'This pill is in use';

    let wrapper;

    afterEach(() => {
        jest.resetAllMocks();
    });

    describe('in inactive, collapsed state (with minimal props)', () => {
        beforeEach(() => {
            wrapper = mount(
                <PillButton toggleDropdown={toggleDropdownMock} onClear={onClearMock} name={name} />
            );
        });

        it('should render correctly', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('.PillButton__pill--isActive')).toHaveLength(0);
            expect(wrapper.find('.PillButton__pill--isOpen')).toHaveLength(0);
        });
        it('should trigger toggle state once when clicked', () => {
            wrapper.find('.PillButton__pill').simulate('click');
            expect(toggleDropdownMock).toHaveBeenCalledTimes(1);
        });
        it('should trigger toggle state once on keyboard interaction', () => {
            wrapper
                .find('.PillButton__pill')
                .simulate('focus')
                .simulate('keyDown', { key: ENTER_KEY });
            expect(toggleDropdownMock).toHaveBeenCalledTimes(1);
        });
        it('should have arrow down label', () => {
            expect(wrapper.find('IoIosArrowDown')).toHaveLength(1);
        });
        it('should trigger toggle state once when button is clicked', () => {
            wrapper.find('button').simulate('click');
            expect(toggleDropdownMock).toHaveBeenCalledTimes(1);
        });
        it('should trigger toggle state once on keyboard interaction with button', () => {
            wrapper.find('button').simulate('focus').simulate('keyDown', { key: ENTER_KEY });
            expect(toggleDropdownMock).toHaveBeenCalledTimes(1);
        });
    });

    describe('in inactive, open state (isOpen prop)', () => {
        beforeEach(() => {
            wrapper = mount(
                <PillButton
                    toggleDropdown={toggleDropdownMock}
                    onClear={onClearMock}
                    name={name}
                    isOpen
                />
            );
        });

        it('should render correctly', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('.PillButton__pill--isActive')).toHaveLength(0);
            expect(wrapper.find('.PillButton__pill--isOpen')).toHaveLength(1);
        });
        it('should have arrow up label', () => {
            expect(wrapper.find('IoIosArrowDown.PillButton__arrowIcon--isOpen')).toHaveLength(1);
        });
        it('should trigger toggle state once when button is clicked', () => {
            wrapper.find('button').simulate('click');
            expect(toggleDropdownMock).toHaveBeenCalledTimes(1);
        });
        it('should trigger toggle state once on keyboard interaction with button', () => {
            wrapper.find('button').simulate('focus').simulate('keyDown', { key: ENTER_KEY });
            expect(toggleDropdownMock).toHaveBeenCalledTimes(1);
        });
    });

    describe('in active, collapsed state (with content prop)', () => {
        beforeEach(() => {
            wrapper = mount(
                <PillButton
                    toggleDropdown={toggleDropdownMock}
                    onClear={onClearMock}
                    name={name}
                    content={content}
                />
            );
        });

        it('should render correctly', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('.PillButton__pill--isActive')).toHaveLength(1);
            expect(wrapper.find('.PillButton__pill--isOpen')).toHaveLength(0);
        });
        it('should not trigger toggle state but onClear only when button is clicked', () => {
            wrapper.find('.PillButton__button').simulate('click');
            expect(onClearMock).toHaveBeenCalledTimes(1);
            expect(toggleDropdownMock).toHaveBeenCalledTimes(0);
        });
        it('should not trigger toggle state but onClear only on keyboard interaction with button', () => {
            wrapper.find('button').simulate('focus').simulate('keyDown', { key: ENTER_KEY });
            expect(onClearMock).toHaveBeenCalledTimes(1);
            expect(toggleDropdownMock).toHaveBeenCalledTimes(0);
        });
    });
    describe('in active, open state (content and isOpen prop)', () => {
        beforeEach(() => {
            wrapper = mount(
                <PillButton
                    toggleDropdown={toggleDropdownMock}
                    onClear={onClearMock}
                    name={name}
                    content={content}
                    isOpen
                />
            );
        });

        it('should render correctly', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('.PillButton__pill--isActive')).toHaveLength(1);
            expect(wrapper.find('.PillButton__pill--isOpen')).toHaveLength(1);
        });
        it('should have arrow up label', () => {
            expect(wrapper.find('IoIosArrowDown.PillButton__arrowIcon--isOpen')).toHaveLength(1);
        });
        it('should trigger toggle state once when button is clicked', () => {
            wrapper.find('button').simulate('click');
            expect(toggleDropdownMock).toHaveBeenCalledTimes(1);
        });
        it('should trigger toggle state once on keyboard interaction with button', () => {
            wrapper.find('button').simulate('focus').simulate('keyDown', { key: ENTER_KEY });
            expect(toggleDropdownMock).toHaveBeenCalledTimes(1);
        });
    });
});
