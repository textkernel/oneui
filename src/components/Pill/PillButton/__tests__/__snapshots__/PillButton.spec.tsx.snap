// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PillButton> component in active, collapsed state (with content prop) should render correctly 1`] = `
<PillButton
  content="This pill is in use"
  name="Pill name"
  onClear={[MockFunction]}
  toggleDropdown={[MockFunction]}
>
  <div
    className="PillButton"
  >
    <div
      className="PillButton__label PillButton__label--isActive"
    >
      Pill name
    </div>
    <div
      className="PillButton__pill PillButton__pill--isActive"
      onClick={[MockFunction]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="PillButton__pillLabel"
      >
        This pill is in use
      </span>
      <button
        className="PillButton__button"
        onClick={[Function]}
        onKeyDown={[Function]}
        type="button"
      >
        <IoMdClose>
          <IconBase
            attr={
              Object {
                "viewBox": "0 0 512 512",
              }
            }
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"
                key="0"
              />
            </svg>
          </IconBase>
        </IoMdClose>
      </button>
    </div>
  </div>
</PillButton>
`;

exports[`<PillButton> component in active, open state (content and isOpen prop) should render correctly 1`] = `
<PillButton
  content="This pill is in use"
  isOpen={true}
  name="Pill name"
  onClear={[MockFunction]}
  toggleDropdown={[MockFunction]}
>
  <div
    className="PillButton"
  >
    <div
      className="PillButton__label PillButton__label--isActive"
    >
      Pill name
    </div>
    <div
      className="PillButton__pill PillButton__pill--isOpen PillButton__pill--isActive"
      onClick={[MockFunction]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="PillButton__pillLabel"
      >
        This pill is in use
      </span>
      <button
        className="PillButton__button"
        type="button"
      >
        <IoIosArrowDown
          className="PillButton__arrowIcon PillButton__arrowIcon--isOpen"
        >
          <IconBase
            attr={
              Object {
                "viewBox": "0 0 512 512",
              }
            }
            className="PillButton__arrowIcon PillButton__arrowIcon--isOpen"
          >
            <svg
              className="PillButton__arrowIcon PillButton__arrowIcon--isOpen"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
                key="0"
              />
            </svg>
          </IconBase>
        </IoIosArrowDown>
      </button>
    </div>
  </div>
</PillButton>
`;

exports[`<PillButton> component in inactive, collapsed state (with minimal props) should render correctly 1`] = `
<PillButton
  name="Pill name"
  onClear={[MockFunction]}
  toggleDropdown={[MockFunction]}
>
  <div
    className="PillButton"
  >
    <div
      className="PillButton__label"
    />
    <div
      className="PillButton__pill"
      onClick={[MockFunction]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="PillButton__pillLabel"
      >
        Pill name
      </span>
      <button
        className="PillButton__button"
        type="button"
      >
        <IoIosArrowDown
          className="PillButton__arrowIcon"
        >
          <IconBase
            attr={
              Object {
                "viewBox": "0 0 512 512",
              }
            }
            className="PillButton__arrowIcon"
          >
            <svg
              className="PillButton__arrowIcon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
                key="0"
              />
            </svg>
          </IconBase>
        </IoIosArrowDown>
      </button>
    </div>
  </div>
</PillButton>
`;

exports[`<PillButton> component in inactive, open state (isOpen prop) should render correctly 1`] = `
<PillButton
  isOpen={true}
  name="Pill name"
  onClear={[MockFunction]}
  toggleDropdown={[MockFunction]}
>
  <div
    className="PillButton"
  >
    <div
      className="PillButton__label"
    />
    <div
      className="PillButton__pill PillButton__pill--isOpen"
      onClick={[MockFunction]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="PillButton__pillLabel"
      >
        Pill name
      </span>
      <button
        className="PillButton__button"
        type="button"
      >
        <IoIosArrowDown
          className="PillButton__arrowIcon PillButton__arrowIcon--isOpen"
        >
          <IconBase
            attr={
              Object {
                "viewBox": "0 0 512 512",
              }
            }
            className="PillButton__arrowIcon PillButton__arrowIcon--isOpen"
          >
            <svg
              className="PillButton__arrowIcon PillButton__arrowIcon--isOpen"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
                key="0"
              />
            </svg>
          </IconBase>
        </IoIosArrowDown>
      </button>
    </div>
  </div>
</PillButton>
`;
